File: src/main/java/rag/app/stages/RetrievalStage.java
package rag.app.stages;

import rag.app.Context;
import rag.app.StrategyRegistry;
import rag.retrieval.Document;
import rag.retrieval.Hit;
import rag.retrieval.Retriever;
import rag.trace.TraceBus;
import rag.trace.TraceEvent;

import java.util.List;

public class RetrievalStage implements PipelineStage {

    private final Retriever retriever;
    private final java.util.List<Document> documents;

    public RetrievalStage(StrategyRegistry registry) {
        this.retriever = registry.getRetriever();
        this.documents = registry.getDocuments();
    }

    @Override
    public String getName() {
        return "RetrievalStage";
    }

    @Override
    public void run(Context context, TraceBus traceBus) throws Exception {

        long start = System.currentTimeMillis();
        String error = null;
        List<Hit> hits = null;
        Exception failure = null;

        try {
            hits = retriever.retrieve(
                    context.getQuestion(),
                    context.getTerms(),
                    documents
            );
            context.setHits(hits);

        } catch (Exception ex) {
            error = ex.getMessage();
            failure = ex;
        } finally {
            long duration = System.currentTimeMillis() - start;

            traceBus.publish(
                    new TraceEvent(
                            getName(),
                            "hits=" + (hits != null ? hits.size() : 0),
                            duration,
                            error
                    )
            );
        }

        if (failure != null) {
            throw failure;
        }
    }
}

File: src/main/python/rag/app/stages/retrieval_stage.py
from typing import List, Optional
from rag.app.stages.pipeline_stage import PipelineStage
from rag.app.context import Context
from rag.app.strategy_registry import StrategyRegistry
from rag.retrieval.document import Document
from rag.retrieval.hit import Hit
from rag.retrieval.retriever import Retriever
from rag.trace.trace_bus import TraceBus
from rag.trace.trace_event import TraceEvent
from rag.utils.timer_utils import now


class RetrievalStage(PipelineStage):
    def __init__(self, registry: StrategyRegistry) -> None:
        self.retriever: Retriever = registry.get_retriever()
        self.documents: List[Document] = registry.get_documents()

    def get_name(self) -> str:
        return "RetrievalStage"

    def run(self, context: Context, trace_bus: TraceBus) -> None:
        start: int = now()
        error: Optional[str] = None
        hits: Optional[List[Hit]] = None
        failure: Optional[Exception] = None

        try:
            hits = self.retriever.retrieve(
                context.get_question(),
                context.get_terms() or [],
                self.documents
            )
            context.set_hits(hits)
        except Exception as ex:
            error = str(ex)
            failure = ex
        finally:
            duration: int = now() - start
            trace_bus.publish(
                TraceEvent(
                    self.get_name(),
                    f"hits={len(hits) if hits else 0}",
                    duration,
                    error
                )
            )

        if failure is not None:
            raise failure






